src/main/java/net/minestom/server/inventory/click/ClickResult.java

```diff
@@ -0,0 +1,165 @@
+        private final @NotNull Player player;
+        private final @NotNull Inventory clickedInventory;
+
+        private final Int2ObjectMap<ItemStack> changes = new Int2ObjectArrayMap<>();
+        private final Int2ObjectMap<ItemStack> playerInventoryChanges = new Int2ObjectArrayMap<>();
+        private @Nullable ItemStack newCursorItem;
+        private @Nullable SideEffects sideEffects;
+
+        Builder(@NotNull Player player, @NotNull Inventory clickedInventory) {
+            this.player = player;
+            this.clickedInventory = clickedInventory;
+        }
+
+        public @NotNull Player player() {
+            return player;
+        }
+
+        public @NotNull Inventory clickedInventory() {
+            return clickedInventory;
+        }
+
+        public @NotNull PlayerInventory playerInventory() {
+            return player().getInventory();
+        }
+
+        public @NotNull ItemStack getCursorItem() {
+            return player().getInventory().getCursorItem();
+        }
+
+        @Override
+        public @NotNull ItemStack get(int slot) {
+            if (slot >= clickedInventory.getSize()) {
+                int converted = PlayerInventoryUtils.protocolToMinestom(slot, clickedInventory);
+
+                return playerInventoryChanges.containsKey(converted) ?
+                        playerInventoryChanges.get(converted) : playerInventory().getItemStack(converted);
+            } else {
+                return changes.containsKey(slot) ?
+                        changes.get(slot) : clickedInventory.getItemStack(slot);
+            }
+        }
+
+        @Override
+        public ItemStack put(int key, ItemStack value) {
+            ItemStack get = get(key);
+            change(key, value);
+            return get;
+        }
+
+        public @NotNull Builder change(int slot, @NotNull ItemStack item) {
+            if (slot >= clickedInventory.getSize()) {
+                change(PlayerInventoryUtils.protocolToMinestom(slot, clickedInventory), item, true);
+            } else {
+                change(slot, item, false);
+            }
+            return this;
+        }
+
+        public @NotNull Builder change(int slot, @NotNull ItemStack item, boolean playerInventory) {
+            (playerInventory ? playerInventoryChanges : changes).put(slot, item);
+            return this;
+        }
+
+        public @NotNull Builder cursor(@Nullable ItemStack newCursorItem) {
+            this.newCursorItem = newCursorItem;
+            return this;
+        }
+
+        public @NotNull Builder sideEffects(@Nullable SideEffects sideEffects) {
+            this.sideEffects = sideEffects;
+            return this;
+        }
+
+        public @NotNull ClickResult build() {
+            return new ClickResult(
+                    player, clickedInventory,
+                    Int2ObjectMaps.unmodifiable(new Int2ObjectArrayMap<>(changes)),
```

Comment:
This should be part of `ClickResult` canonical construtor, not dependent on the builder